File: library/src/main/java/com/lihb/library/CustomViewPager.java
package com.lihb.library;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.MotionEvent;

import androidx.viewpager.widget.ViewPager;

/**
 * 可以设置是否响应手势切换的viewpager
 * Created by lihuabin on 2016/10/10.
 */
public class CustomViewPager extends ViewPager {

    private boolean isCanScroll = true;

    public CustomViewPager(Context context) {
        super(context);
    }

    public CustomViewPager(Context context, AttributeSet attrs) {
        super(context, attrs);
        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.CustomViewPager, 0, 0);
        isCanScroll = a.getBoolean(R.styleable.CustomViewPager_canScroll, true);
        a.recycle();
    }

    public boolean isCanScroll() {
        return isCanScroll;
    }

    public void setCanScroll(boolean isCanScroll) {
        this.isCanScroll = isCanScroll;
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        return this.isCanScroll && super.onTouchEvent(event);
    }

    @Override
    public boolean onInterceptTouchEvent(MotionEvent event) {
        return this.isCanScroll && super.onInterceptTouchEvent(event);
    }
}

File: library/src/main/java/com/lihb/customviewpager/CustomViewPager.kt
package com.lihb.customviewpager

import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import androidx.viewpager.widget.ViewPager

/**
 * 可以设置是否响应手势切换的viewpager
 * Created by lihuabin on 2016/10/10.
 */
open class CustomViewPager
@JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null
) : ViewPager(context, attrs) {
    var isCanScroll = true

    init {
        val a = context.obtainStyledAttributes(attrs, R.styleable.CustomViewPager, 0, 0)
        isCanScroll = a.getBoolean(R.styleable.CustomViewPager_canScroll, true)
        a.recycle()
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        return isCanScroll && super.onTouchEvent(event)
    }

    override fun onInterceptTouchEvent(event: MotionEvent): Boolean {
        return isCanScroll && super.onInterceptTouchEvent(event)
    }

    fun doOnPageSelected(action: (Int) -> Unit): OnPageChangeListener {
        val listener = object : OnPageChangeListener {
            override fun onPageScrollStateChanged(state: Int) {

            }

            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
            }

            override fun onPageSelected(position: Int) {
                action.invoke(position)
            }

        }
        addOnPageChangeListener(listener)
        return listener
    }


}